p2p/kademlia/store/sqlite/sqlite.go

```diff
@@ -0,0 +1,201 @@
+type Store struct {
+	db                *sqlx.DB
+	replicateInterval time.Duration
+	republishInterval time.Duration
+}
+
+// Record is a data record
+type Record struct {
+	Key           string
+	Data          []byte
+	IsOriginal    bool
+	CreatedAt     time.Time
+	UpdatedAt     time.Time
+	ReplicatedAt  time.Time
+	RepublishedAt time.Time
+}
+
+// NewStore returns a new store
+func NewStore(ctx context.Context, dataDir string, replicate time.Duration, republish time.Duration) (*Store, error) {
+	s := &Store{}
+
+	log.P2P().WithContext(ctx).Debugf("p2p data dir: %v", dataDir)
+	if _, err := os.Stat(dataDir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dataDir, 0750); err != nil {
+			return nil, fmt.Errorf("mkdir %q: %w", dataDir, err)
+		}
+	} else if err != nil {
+		return nil, fmt.Errorf("cannot create data folder: %w", err)
+	}
+
+	dbFile := path.Join(dataDir, "foo.db")
+	db, err := sqlx.Connect("sqlite3", dbFile)
+	if err != nil {
+		return nil, fmt.Errorf("cannot open sqlite database: %w", err)
+	}
+
+	s.db = db
+	s.replicateInterval = replicate
+	s.republishInterval = republish
+
+	if !s.checkStore() {
+		if err = s.migrate(); err != nil {
+			return nil, fmt.Errorf("cannot create table(s) in sqlite database: %w", err)
+		}
+	}
+
+	return s, nil
+}
+
+func (s *Store) checkStore() bool {
+	query := `SELECT name FROM sqlite_master WHERE type='table' AND name='data'`
+	var name string
+	err := s.db.Get(&name, query)
+	return err == nil
+}
+
+func (s *Store) migrate() error {
+	query := `
+    CREATE TABLE IF NOT EXISTS data(
+        key TEXT PRIMARY KEY,
+        data BLOB NOT NULL,
+        is_original BOOL DEFAULT FALSE,
+        createdAt DATETIME DEFAULT CURRENT_TIMESTAMP,
+        updatedAt DATETIME DEFAULT CURRENT_TIMESTAMP,
+        replicatedAt DATETIME,
+        republishedAt DATETIME
+    );
+    `
+
+	if _, err := s.db.Exec(query); err != nil {
+		return fmt.Errorf("failed to create table 'data': %w", err)
+	}
+	return nil
+}
+
+// Store will store a key/value pair for the local node
+func (s *Store) Store(_ context.Context, key []byte, value []byte) error {
+	hkey := hex.EncodeToString(key)
+
+	now := time.Now().UTC()
+	r := Record{Key: hkey, Data: value, UpdatedAt: now}
+	res, err := s.db.NamedExec(`INSERT INTO data(key, data) values(:key, :data) ON CONFLICT(key) DO UPDATE SET updatedAt=:updatedat`, r)
+	if err != nil {
+		return fmt.Errorf("cannot insert or update record with key %s: %w", hkey, err)
+	}
+
+	if rowsAffected, err := res.RowsAffected(); err != nil {
+		return fmt.Errorf("failed to insert/update record with key %s: %w", hkey, err)
+	} else if rowsAffected == 0 {
+		return fmt.Errorf("failed to insert/update record with key %s", hkey)
+	}
+
+	return nil
+}
+
+// Retrieve will return the local key/value if it exists
+func (s *Store) Retrieve(_ context.Context, key []byte) ([]byte, error) {
+	hkey := hex.EncodeToString(key)
+
+	r := Record{}
+	err := s.db.Get(&r, `SELECT data FROM data WHERE key LIKE ?`, hkey)
```

Comment:
shouldn't we use = here instead of LIKE?
LIKE is usually very expensive

Reply:
`=` didn't work at all

Reply:
although, it didn't work when i was writing key as is, now i convert it to hex string, i will re-test

Reply:
ok - I've approved just in case